consensus/consensus-types/src/common.rs

```diff
@@ -64,18 +123,28 @@ impl Payload {
                 .with_min_len(100)
                 .map(|txn| txn.raw_txn_bytes_len())
                 .sum(),
+            Payload::InQuorumStore(proof_with_status) => proof_with_status
+                .proofs
+                .iter()
+                .map(|proof| proof.info().num_bytes as usize)
+                .sum(),
         }
     }
-}
-
-// TODO: What I really want is an iterator that isn't necessarily a vector (e.g., read lazily from RocksDB). This doesn't seem like the way.
-impl IntoIterator for Payload {
-    type Item = SignedTransaction;
-    type IntoIter = std::vec::IntoIter<Self::Item>;
 
-    fn into_iter(self) -> Self::IntoIter {
-        match self {
-            Payload::DirectMempool(txns) => txns.into_iter(),
+    pub fn verify(
+        &self,
+        validator: &ValidatorVerifier,
+        quorum_store_enabled: bool,
+    ) -> anyhow::Result<()> {
+        match (quorum_store_enabled, self) {
+            (false, Payload::DirectMempool(_)) => Ok(()),
+            (true, Payload::InQuorumStore(proof_with_status)) => {
+                for proof in proof_with_status.proofs.iter() {
+                    proof.verify(validator)?;
+                }
+                Ok(())
+            }
+            (_, _) => Err(VerifyError::WrongPayload).context("Failed to verify Payload"),
```

Comment:
Do we need the type? Just use `Err(anyhow::anyhow!(...)`? Also maybe we can add the expected and actual payload types to the message?